### DISCUSSION TITLE
🛠️ Discussion on Tool support

### DISCUSSION THREAD

**Discussion by anakin87** [deepset-ai/haystack-experimental#98] [General]
**UPDATE: All work on Tool has been incorporated into Haystack 2.9.0. See the [docs](https://docs.haystack.deepset.ai/docs/tool)**

---
This is the discussion board for a group of experimental components and dataclasses to improve the support for Tool calling in Haystack.

**What is Tool calling/function calling?**
A growing number of Language Models now support passing tool definitions alongside the prompt.
Tools can be functions or APIs.
When appropriate, the model can generate a call to the tool in order to address the user’s query, _although the model itself does not execute the tool_.

The features outlined below are built to support tools in Haystack in a unified way, covering the entire process from tool call preparation to proper invocation and potential post-processing.

**Experimental features:**
- [refactored `ChatMessage` dataclass](https://github.com/deepset-ai/haystack-experimental/blob/main/haystack_experimental/dataclasses/chat_message.py) to be more flexible and future-proof.
- [`Tool` dataclass](https://github.com/deepset-ai/haystack-experimental/tree/main/haystack_experimental/dataclasses/tool.py): a simple and unified abstraction to represent Tools throughout the framework.
- [refactored `OpenAIChatGenerator`](https://github.com/deepset-ai/haystack-experimental/blob/main/haystack_experimental/components/generators/chat/openai.py), with support for tools.
- [`ToolInvoker` component](https://github.com/deepset-ai/haystack-experimental/blob/main/haystack_experimental/components/tools/tool_invoker.py), to actually execute tool calls prepared by Language Models.

🚧 We are working on adding support for more Chat Generators and developing easier ways to convert functions into tools...

🧑‍🍳 **[Try the Cookbook here](https://colab.research.google.com/github/deepset-ai/haystack-cookbook/blob/main/notebooks/tools_support.ipynb)**. Contains a step by step guide to the new features and shows several end-to-end examples.

🧑‍🍳 New recipe: [Newsletter Sending Agent](https://haystack.deepset.ai/cookbook/newsletter-agent)

Let us know how you get along with these components, and please do let us know if you have any feedback.

> **Note:** Experimental features live in this repository for a fixed period of time. We don't guarantee that we will continue maintaining experimental features. But if they are successful, stable and if you like it, we will move the feature to the core Haystack package.

**Comment by cjayb** [comment]
Very much appreciate this abstraction! Makes working with, and especially debugging, tool calls a lot more transparent. I need to use Azure OpenAI, but managed to port #57 over (see my comment to #57 if you'd like me to contribute it back?)

**Comment by anakin87** [comment]
@cjayb Thanks for the appreciation!

Currently, this feature is experimental and we want to support it in 3 Chat Generators: OpenAI, Antrophic, and Ollama.
We will use the experiment time to iterate quickly and get feedback.

If the experiment is successful, starting in November, we will implement it for other Generators as well and I can ping you to contribute.

**Reply by cjayb** [reply]
Sounds like a plan 👏

**Comment by anakin87** [comment]
During the live, a community member mentioned that he would like to use the  `@tool` decorator in addition to the class method `Tool.from_function`

**Comment by anakin87** [comment]
🧑‍🍳 New recipe: [Newsletter Sending Agent with Experimental Haystack Tools](https://haystack.deepset.ai/cookbook/newsletter-agent)

**Comment by anakin87** [comment]
We also added support for Tools in Ollama, Anthropic and Hugging Face API Chat Generators!

**Comment by Emil-io** [comment]
**Idea: Output of Tool Invoker matches input of ChatGenerator**

The proper workflow when using tools requires 2 generations: 
(1) The LLM outputs that a function call should be performed
(2) Based on the chat history + Last LLM reply + Tool Result a new generation is started, which outputs a response text message

For this to properly work, I have two options
(a) Use two separate Generator components in my pipeline. First generator component is for Step (1), after that there is a router, which routes the output to Generator 2 for Step (2) , if in Step (1) a tool call was requested _(Not preferred)_
(b) Use the same generator with a loop. Easiest would be, if the function call output of Generator (1) could simply be passed to the Tool Invoker. And the the output of the tool invoker can then directly be provided to Generator (2) for the generation of text

Since (b) is preferred:
This would mean, that the Tool Invoker outputs the Chat History + Last LLM reply + Tool Result.
I think this would make the workflow for working with tools very smooth (only one routing after Generator (1) and then a very simple loop by directly connecting the Tool Invoker to the Generator again)

**Alternatively**, a new component called OpenAIAgent, which for Step (2) automatically processes the output of the tool invoker in its current type, and then also turns it into the appropriate chat message list based on the chat messages it was provided with for first generation

P.S. Maybe routing could even be performed by the Generator, since this follows a fixed syntax (either for ChatGeneraotr or OpenAIAgent :) )

**Reply by Emil-io** [reply]
@anakin87 wdyt?

**Reply by anakin87** [reply]
I will have a proper look this week

**Comment by farzbood** [comment]
Hi,
I've tried the [cookbook](https://haystack.deepset.ai/cookbook/swarm) code with  a number of **HuggingFace** open models (`HuggingFaceH4/starchat2-15b-v0.1` , `mistralai/Mistral-7B-Instruct-v0.3`, `Qwen/QwQ-32B-Preview`) , it doesn't work as expected with any of them. (the conversation stuck with `joker` with a response says the request can't be processed and back to the **User prompt**, in the best case scenario, dialog transferred to` Refund Agent` but still says it can't process the refund request)
Is this type of collaborative agent model depends on a specific LLM capability or provider?

**Comment by anakin87** [comment]
**UPDATE: All work on Tool has been incorporated into Haystack 2.9.0. See the [docs](https://docs.haystack.deepset.ai/docs/tool)**.
